upstairs/src/volume.rs

```diff
@@ -3490,55 +3510,116 @@ mod test {
         // A valid replacement VCR is provided with only one target being
         // different.
         // Test all three targets for replacement.
-        for cid in 0..3 {
-            test_volume_replace_cid(cid).unwrap();
+        // Test with 1, 2, and 3 sub_volumes.
+        // Test with the difference being in each sub_volume.
+        // Test both with and without a read_only_parent.
+        for sv in 1..4 {
+            for sv_changed in 0..sv {
+                for cid in 0..3 {
+                    test_volume_replace_inner(sv, sv_changed, cid, false)
+                        .unwrap();
+                    test_volume_replace_inner(sv, sv_changed, cid, true)
+                        .unwrap();
+                }
+            }
         }
     }
 
-    fn test_volume_replace_cid(cid: usize) -> Result<()> {
+    // A basic replacement of a target in a sub_volume, with options to
+    // create multiple sub_volumes, and to select which sub_volume and specific
+    // target to be different.
+    //
+    // sv_count: The number of sub volumes we create in each volume.
+    // sv_changed: The index (zero based) of the sub volume where we
+    // want the target to be different.
+    // cid: The client ID where the change will happen, which is the
+    // same as the index in the crucible opts target array.
+    // add_rop: True to create a read_only_parent, false for None
+    fn test_volume_replace_inner(
+        sv_count: usize,
+        sv_changed: usize,
+        cid: usize,
+        add_rop: bool,
+    ) -> Result<()> {
+        assert!(sv_count > sv_changed);
+        assert!(cid < 3);
         // A valid replacement VCR is provided with a larger generation
         // number and only one target being different.
+        let vol_id = Uuid::new_v4();
+        let opts = generic_crucible_opts(vol_id);
+
         let block_size = 512;
         let vol_id = Uuid::new_v4();
         let blocks_per_extent = 10;
         let extent_count = 9;
 
-        let mut opts = generic_crucible_opts(vol_id);
+        let rop = if add_rop {
+            let rop = Box::new(VolumeConstructionRequest::Region {
+                block_size,
+                blocks_per_extent,
+                extent_count,
+                opts: opts.clone(),
+                gen: 3,
+            });
+            Some(rop)
+        } else {
+            None
+        };
 
-        let original = VolumeConstructionRequest::Volume {
-            id: vol_id,
-            block_size,
-            sub_volumes: vec![VolumeConstructionRequest::Region {
+        let mut sub_volumes = Vec::new();
+        for _ in 0..sv_count {
+            let sv = VolumeConstructionRequest::Region {
                 block_size,
                 blocks_per_extent,
                 extent_count,
                 opts: opts.clone(),
                 gen: 2,
-            }],
-            read_only_parent: None,
+            };
+            sub_volumes.push(sv);
+        }
```

Comment:
Nit: this could be slightly shorter + immutable
```rust
        let sub_volumes = (0..sv_count)
            .map(|_| VolumeConstructionRequest::Region {
                block_size,
                blocks_per_extent,
                extent_count,
                opts: opts.clone(),
                gen: 2,
            })
            .collect();
```

Reply:
ab9e0223 has this immutable update, plus I made a specific function and called it wherever I could.